minimize: restructe const evaluation

bifbof opened the pull request:
I changed the constant evaluation according to #158.
I wanted to find a cleaner way to deal with the pointer translation, but in the end kept the old system.

RalfJung commented on tooling/minimize/src/constant.rs:
```diff
@@ -2,87 +2,82 @@ use crate::*;
-                self.translate_const_uneval(uneval, *ty, span),
-            rs::mir::Const::Val(val, ty) => self.translate_const_val(val, *ty, span),
-        }
+        let val = c.eval(self.tcx, rs::ParamEnv::reveal_all(), None).expect("const-eval failed");
```
Would be good to use span_bug when const eval fails

RalfJung commented on tooling/minimize/src/constant.rs:
```diff
@@ -2,87 +2,82 @@ use crate::*;
-                    Signed => val.try_to_int(val.size()).unwrap().into(),
-                    Unsigned => val.try_to_uint(val.size()).unwrap().into(),
+                let scalar = ecx.read_scalar(&val).unwrap();
+                let val = match int_ty.signed {
```
```suggestion
                let val: Int = match int_ty.signed {
```

RalfJung commented on tooling/minimize/src/constant.rs:
```diff
@@ -2,87 +2,82 @@ use crate::*;
+                }
+                let ptr = ecx.read_pointer(&val).unwrap();
+                let (prov, offset) = ptr.into_parts();
+                let alloc_id = prov.unwrap().alloc_id();
```
This can be `None` if the constant is something like
```rust
const C: *const i32 = 4 as *const i32;
```

bifbof replied:
Now create a raw pointer if provenance is `None`.

RalfJung commented on tooling/minimize/src/constant.rs:
```diff
@@ -136,12 +131,10 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
-                    inner_offset_bytes.iter().map(|x| x.unwrap()).collect();
-                let inner_offset: Int =
-                    DefaultTarget::ENDIANNESS.decode(Unsigned, inner_offset_bytes);
+                let start = offset.bytes() as usize;
```
```suggestion
                let start = offset.bytes_usize();
```

RalfJung commented on tooling/minimize/src/constant.rs:
```diff
@@ -136,12 +131,10 @@ impl<'cx, 'tcx> FnCtxt<'cx, 'tcx> {
-                    DefaultTarget::ENDIANNESS.decode(Unsigned, inner_offset_bytes);
+                let start = offset.bytes() as usize;
+                let end = start + DefaultTarget::PTR_SIZE.bytes().try_to_usize().unwrap();
+                let inner_offset = bytes[start..end].iter().map(|x| x.unwrap()).collect();
```
```suggestion
                // Pointer bytes are always initialized, so we can unwrap.
                let inner_offset = bytes[start..end].iter().map(|x| x.unwrap()).collect();
```

RalfJung reviewed (commented):
Thanks, looks good overall! However, please add a test file that ensures we can correct translate a bunch of constants (at least one for each case in translate_const_val).

bifbof commented:
Thanks for the review! I have added the requested changes @RalfJung.
There were also some additional changes where I fixed a bug where I was using the `VariantIdx` instead of the real discriminator in the enums.

RalfJung commented:
> There were also some additional changes where I fixed a bug where I was using the VariantIdx instead of the real discriminator in the enums.

Is that covered by the tests you added?

bifbof commented:
> > There were also some additional changes where I fixed a bug where I was using the VariantIdx instead of the real discriminator in the enums.
> 
> Is that covered by the tests you added?

Yes, the  `assert!(!(() > ()));` and the assert with `NEG_ONE` and `NEG_TWO` test for it.

RalfJung commented:
Great, thanks!
